go/adbc/driver/snowflake: GetObjects panics while retrieving table columns of vector type

### What happened?

When a table contains a column of VECTOR data type, the driver panics:
`panic: runtime error: invalid memory address or nil pointer dereference`

To my understanding, this is not a [supported type](https://arrow.apache.org/adbc/current/driver/snowflake.html#type-support) by this driver or Go clients in general.

However, the driver could return a clearer error message.
 

### Stack Trace

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x103f614c0]

goroutine 74 [running]:
github.com/apache/arrow-adbc/go/adbc/driver/internal/driverbase.ToXdbcDataType({0x0, 0x0})
        /Users/<redacted>/go/pkg/mod/github.com/apache/arrow-adbc/go/adbc@v1.4.0/driver/internal/driverbase/connection.go:703 +0x20
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.(*connectionImpl).GetObjects.func4()
        /Users/<redacted>/go/pkg/mod/github.com/apache/arrow-adbc/go/adbc@v1.4.0/driver/snowflake/connection.go:368 +0x280
created by github.com/apache/arrow-adbc/go/adbc/driver/snowflake.(*connectionImpl).GetObjects in goroutine 1
        /Users/<redacted>/go/pkg/mod/github.com/apache/arrow-adbc/go/adbc@v1.4.0/driver/snowflake/connection.go:354 +0x164c
exit status 2
```

### How can we reproduce the bug?

- Create the table described in this doc: https://docs.snowflake.com/en/sql-reference/data-types-vector#loading-and-unloading-vector-data
- Send a `GetObjects` at `ObjectDepthColumns` targetting that table

### Environment/Setup

_No response_

--- a/go/adbc/driver/internal/shared_utils.go
+++ b/go/adbc/driver/internal/shared_utils.go
@@ -694,6 +694,10 @@ const (
 )
 
 func ToXdbcDataType(dt arrow.DataType) (xdbcType XdbcDataType) {
+	if dt == nil {
+		return XdbcDataType_XDBC_UNKNOWN_TYPE
+	}
+
 	switch dt.ID() {
 	case arrow.EXTENSION:
 		return ToXdbcDataType(dt.(arrow.ExtensionType).StorageType())

--- a/go/adbc/driver/snowflake/connection.go
+++ b/go/adbc/driver/snowflake/connection.go
@@ -350,7 +350,9 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 						field := c.toArrowField(col)
 						xdbcDataType := internal.ToXdbcDataType(field.Type)
 
-						getObjectsCatalog.CatalogDbSchemas[i].DbSchemaTables[j].TableColumns[k].XdbcDataType = driverbase.Nullable(int16(field.Type.ID()))
+						if field.Type != nil {
+							getObjectsCatalog.CatalogDbSchemas[i].DbSchemaTables[j].TableColumns[k].XdbcDataType = driverbase.Nullable(int16(field.Type.ID()))
+						}
 						getObjectsCatalog.CatalogDbSchemas[i].DbSchemaTables[j].TableColumns[k].XdbcSqlDataType = driverbase.Nullable(int16(xdbcDataType))
 					}
 				}
@@ -475,6 +477,11 @@ func (c *connectionImpl) toArrowField(columnInfo driverbase.ColumnInfo) arrow.Fi
 		fallthrough
 	case "GEOMETRY":
 		field.Type = arrow.BinaryTypes.String
+	case "VECTOR":
+		// despite the fact that Snowflake *does* support returning data
+		// for VECTOR typed columns as Arrow FixedSizeLists, there's no way
+		// currently to retrieve enough metadata to construct the proper type
+		// for it
 	}
 
 	return field

--- a/go/adbc/driver/snowflake/driver_test.go
+++ b/go/adbc/driver/snowflake/driver_test.go
@@ -2294,3 +2294,70 @@ ORDER BY start_time;
 	suite.False(rdr.Next())
 	suite.Require().NoError(rdr.Err())
 }
+
+func (suite *SnowflakeTests) TestGetObjectsVector() {
+	suite.Require().NoError(suite.Quirks.DropTable(suite.cnxn, "MYVECTORTABLE"))
+	suite.Require().NoError(suite.stmt.SetSqlQuery(`CREATE OR REPLACE TABLE myvectortable (
+		a VECTOR(float, 3), b VECTOR(float, 3))`))
+	_, err := suite.stmt.ExecuteUpdate(suite.ctx)
+	suite.Require().NoError(err)
+	suite.Require().NoError(suite.stmt.SetSqlQuery(`INSERT INTO myvectortable
+		SELECT [1.1,2.2,3]::VECTOR(FLOAT,3), [1,1,1]::VECTOR(FLOAT,3)`))
+	_, err = suite.stmt.ExecuteUpdate(suite.ctx)
+	suite.Require().NoError(err)
+
+	tableName := "MYVECTORTABLE"
+	rdr, err := suite.cnxn.GetObjects(suite.ctx, adbc.ObjectDepthColumns, nil, nil, &tableName, nil, nil)
+	suite.Require().NoError(err)
+	defer rdr.Release()
+
+	suite.Require().True(rdr.Next())
+	rec := rdr.Record()
+
+	for i := 0; int64(i) < rec.NumRows(); i++ {
+		// list<db_schema_schema>
+		dbSchemasList := rec.Column(1).(*array.List)
+		// db_schema_schema (struct)
+		dbSchemas := dbSchemasList.ListValues().(*array.Struct)
+		// list<table_schema>
+		dbSchemaTablesList := dbSchemas.Field(1).(*array.List)
+		// table_schema (struct)
+		dbSchemaTables := dbSchemaTablesList.ListValues().(*array.Struct)
+		// list<column_schema>
+		tableColumnsList := dbSchemaTables.Field(2).(*array.List)
+		// column_schema (struct)
+		tableColumns := tableColumnsList.ListValues().(*array.Struct)
+
+		start, end := dbSchemasList.ValueOffsets(i)
+		for j := start; j < end; j++ {
+			schemaName := dbSchemas.Field(0).(*array.String).Value(int(j))
+			if !strings.EqualFold(schemaName, suite.Quirks.DBSchema()) {
+				continue
+			}
+			tblStart, tblEnd := dbSchemaTablesList.ValueOffsets(int(j))
+			for k := tblStart; k < tblEnd; k++ {
+				tblName := dbSchemaTables.Field(0).(*array.String).Value(int(k))
+				if !strings.EqualFold(tblName, tableName) {
+					continue
+				}
+
+				colStart, colEnd := tableColumnsList.ValueOffsets(int(k))
+				suite.EqualValues(2, colEnd-colStart)
+
+				for l := colStart; l < colEnd; l++ {
+					colName := tableColumns.Field(0).(*array.String).Value(int(l))
+					ordinalPos := tableColumns.Field(1).(*array.Int32).Value(int(l))
+					typeName := tableColumns.Field(4).(*array.String).Value(int(l))
+					switch ordinalPos {
+					case 1:
+						suite.Equal("A", colName)
+					case 2:
+						suite.Equal("B", colName)
+					}
+
+					suite.Equal("VECTOR", typeName)
+				}
+			}
+		}
+	}
+}

